### PR TITLE
185847014 annuals due filter

### PR DESCRIPTION
**Pull request by ZachCMP**
## Description

Include a summary of the changes and a link to the related issue. List any dependencies that are required for this change.

## Type of change
- [ ] Bug fix
- [x] New feature (adds functionality)
- [ ] Code clean-up / housekeeping
- [ ] Dependency update

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

### CODE REVIEW

**Comment by ZachCMP** [comment]
Relies on frontend PR https://github.com/greenriver/hmis-frontend/pull/544 (Only schema changes)

@gigxz I took my best shot at porting the logic from the reminder generator to SQL. There was not an obvious way in AREL to to the date transformation done in the reminder generator required to get the enrollment anniversary, so I just used the postgres `make_date` function, but I don't think that'd work in all flavors of SQL. Maybe have Elliot take a look and see if that can be cleaned up.

I copied the logic as-is, but I do have a concern that this scope might not be correct. I think it would indicate that an annual is due if it hasn't had one in 30 days, even if it has had one this year. Maybe it's my misreading of the logic in the generator, or this context is different enough from the generator that it needs to be adjusted. Take a look and let me know what you think. I can make adjustments if needed.

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb [hunk 1]
```diff
@@ -49,4 +50,35 @@ def with_project_types(scope)
   def with_search_term(scope)
     with_filter(scope, :search_term) { scope.matching_search_term(input.search_term) }
   end
+
+  def with_household_tasks(scope)
+    with_filter(scope, :household_tasks) do
+      if input.household_tasks&.include?('ANNUAL_DUE')
+        entry_anniversary = Arel.sql('make_date(extract(year from current_date)::integer, extract(month from "EntryDate")::integer, extract(day from "EntryDate")::integer)')
+        start_date = Arel.sql('make_date(extract(year from current_date)::integer, extract(month from "EntryDate")::integer, extract(day from "EntryDate")::integer) - interval \'30 days\'')
```
Comment (eanders) [inline drivers/hmis/app/models/hmis/filter/enrollment_filter.rb]: This is fine with me.  We could probably re-write it using the `nf` method in ArelHelper, but I don't think it'd be any more clear and it wouldn't be any more portable.  I save leave it as is.

Comment (ttoomey) [inline drivers/hmis/app/models/hmis/filter/enrollment_filter.rb]: This might be a little easier to read with an early return and some here-docs... maybe:
```ruby
  def with_household_tasks(scope)
    return scope.all unless input.household_tasks&.include?('ANNUAL_DUE')

    entry_anniversary = Arel.sql <<~SQL
      make_date(
        extract(year from current_date)::integer,
        extract(month from "EntryDate")::integer,
        extract(day from "EntryDate")::integer
      )
    SQL

    start_date = Arel.sql <<~SQL
      make_date(
        extract(year from current_date)::integer,
        extract(month from "EntryDate")::integer,
        extract(day from "EntryDate")::integer
      ) - interval '30 days'
    SQL
```

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb [hunk 2]
```diff
@@ -49,4 +50,35 @@ def with_project_types(scope)
...
+
+        assessment_ids = Hmis::Hud::CustomAssessment.
+          where(
+            enrollment_id: scope.pluck(:enrollment_id),
+            data_source_id: scope.joins(:project).pluck(p_t[:data_source_id]).uniq,
+            data_collection_stage: 5,
+            wip: false,
+          ).
+          pluck(:id)
```
Comment (ttoomey) [inline drivers/hmis/app/models/hmis/filter/enrollment_filter.rb]: This works but the 'pluck' pattern doesn't scale very well. We can leave this as is and come back to it but I'd suggest we avoid this in the future.

File: db/reporting_structure.sql [hunk 3]
```diff
@@ -22,50 +22,50 @@ SET row_security = off;
 
 CREATE FUNCTION public.monthly_reports_insert_trigger() RETURNS trigger
```
Comment (ttoomey) [inline db/reporting_structure.sql]: There's no migrations in this branch so I don't think the structure files should be included. Suggest:
```
git checkout release-92 db/reporting_structure.sql db/structure.sql db/health_structure.sql db/warehouse_structure.sql
```

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb [hunk 4]
```diff
@@ -49,4 +50,35 @@ def with_project_types(scope)
...
+
+        # SQL-ized version of the logic here: drivers/hmis/app/models/hmis/reminders/reminder_generator.rb#annual_assessment_reminder
+        scope = scope.
+          left_outer_joins(:exit).
+          joins(e_t.join(cas_t, Arel::Nodes::OuterJoin).on(cas_t[:enrollment_id].eq(e_t[:enrollment_id]).and(cas_t[:id]).in(assessment_ids)).join_sources).
+          # Include anything where the entry date is outside of the year window
+          where(e_t[:entry_date].lteq(Date.today - 1.year + 30.days)).
+          # Include anything that isn't exited, or was exited after the entry anniversary
+          where(ex_t[:exit_date].eq(nil).or(ex_t[:exit_date].gt(entry_anniversary))).
```
Comment (ttoomey) [inline drivers/hmis/app/models/hmis/filter/enrollment_filter.rb]: I might be missing something but it seems like this would use the anniversary of the enrollment entry date. But I think it should be the entry date of the HOH, which could be different. Perhaps some inaccuracy is okay?

More broadly, I wonder if there's a way we can avoid this SQL implementation. Maybe we could pre calculate the assessment due dates and then this query becomes simpler. But that's a design question for when Gig returns

Comment (eanders) [reply drivers/hmis/app/models/hmis/filter/enrollment_filter.rb]: I concur, the anniversary date should be x-years from the HoH Entry Date.

Maybe instead of pre-calculating, we could make a helper class that takes some columns, etc. and returns the SQL for the anniversary range.  At least that way we could standardize how these are calculated (I think we need to do that in a handful of places) and if we decided to change it, it would be centralized so should be easy to update as long as we can keep the same API.

Comment (ttoomey) [reply drivers/hmis/app/models/hmis/filter/enrollment_filter.rb]: > the anniversary date should be x-years from the HoH Entry Date.

Per standup convo 11/13, this is actually the earliest entry date for any enrollment in the household, not the current HOH entry date

Comment (eanders) [reply drivers/hmis/app/models/hmis/filter/enrollment_filter.rb]: From the Glossary:
TLDR (calculate from the earliest enrollment start date in the household)
```
Annual Assessment
Reports utilizing Annual Assessment ([data collection stage] = 5) data on stayers in the project 365 days or more require data from the specific Annual Assessment on the head of household’s anniversary most relevant to the [report date range]. The instructions below describe how to select this exact record for each client.
1.	Use the days between the earliest Project Start Date of the household’s project enrollment and the end date of the report when determining length of stay for this purpose. 
2.	In the event a household has more than one head of household active in the report range, (i.e., if the first HoH exited and another household member became the HoH), the later HoH’s [project start date] will be back-dated to the first HoH’s [project start date] and the first HoH’s [relationship to head of household] will be updated to something other than “Self”. This will maintain determining the Annual Assessment timing on the earliest [project start date] for any given household.
3.	Calculate the head of household’s number of years in the project. This can be done using the same algorithm as for calculating a client’s age as of a certain date. Use the client’s [project start date] and the [report end date] as the two dates of comparison. This calculation is time-based, not service-based, so a client with a [project start date] that is active for a year or longer in a night-by-night emergency shelter would require an Annual Assessment, regardless of how many bed night dates were recorded. It is important to use the “age” method of determining client anniversaries due to leap years; using “one year = 365 days” will eventually incorrectly offset the calculated anniversaries of long-term stayers.
4.	If the HoH’s number of years in the project is 0, the household is not yet required to have an annual assessment. 
5.	If the HoH’s number of years in the project is more than 0, add that number to the year of the HoH’s [project start date]. This becomes the household’s relevant anniversary date for the purposes of the report. For example, using a report date range of 10/1/2022 – 9/30/2023, a HoH with a [project start date] of 6/1/2021 will have been in the project 2 years as of 9/30/2023 and so will have an [anniversary date] of 6/1/2023. (i.e. 2021 + 2 = 2023). 
6.	Use the latest annual assessment ([data collection stage] = 5) for each client in the household dated between:
a.	30 days prior to the [anniversary date] (even if this date falls before the [report start date]) 
b.	and the lesser of (30 days after the [anniversary date], [report end date])
7.	If the household member’s [project start date] is > [anniversary date] (i.e., the person entered the household after the annual assessment may have been collected), that person is not yet required to have an annual assessment.
8.	Exclude any data with an [information date] > [report end date], even though it is legitimate for a client’s relevant annual assessment data to fall in this date range for clients whose anniversaries are near to the end of the [report date range].
9.	Clients with no annual assessment data in the relevant date range as indicated in step 6 may be reported as “missing annual assessment” or may be completely omitted depending on the report question.
10.	Clients who aren’t required to have an annual assessment yet as determined in steps 4 and 7 may be reported accordingly or may be completely omitted depending on the question.
```

Comment (ZachCMP) [reply drivers/hmis/app/models/hmis/filter/enrollment_filter.rb]: @eanders @ttoomey I fixed this up to both be cleaner and match the spec better. Take a look and let me know if you'd like any more improvements. I also added tests for this, please let me know if you'd like more test cases in there.

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb (hunk 2 above)
Comment (ZachCMP) [reply drivers/hmis/app/models/hmis/filter/enrollment_filter.rb]: Agreed, I was able to remove this

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb (hunk 4 above)
Comment (eanders) [reply drivers/hmis/app/models/hmis/filter/enrollment_filter.rb]: I think I'm good.  I'll defer to Gig for a thorough review of that, on the surface, it looks correct to me.

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb [hunk 5]
```diff
@@ -13,6 +13,7 @@ def filter_scope(scope)
       yield_self(&method(:with_bed_night_on_date)).
       yield_self(&method(:with_project_types)).
       yield_self(&method(:with_search_term)).
+      yield_self(&method(:with_household_tasks)).
       yield_self(&method(:clean_scope))
```
Comment (gigxz) [inline drivers/hmis/app/models/hmis/filter/enrollment_filter.rb:17]: I'm getting the below error when I try to apply the filter. 2023 wasn't a leap year so 2/29 isn't valid, which I'm assuming the error is coming from. When I look in the db though, I don't have any Enrollments or Assessments with that date. I have lots with dates `2023-03-29`/`2022-03-29`, though. I wonder if the `- 30 days` is causing ending up on the invalid date. Can you repro if you create an enrollment with an entry date of 3/29?

```
PG::DatetimeFieldOverflow: ERROR: date field value out of range: 2023-02-29
```

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb [hunk 6]
```diff
@@ -49,4 +50,52 @@ def with_project_types(scope)
   def with_search_term(scope)
     with_filter(scope, :search_term) { scope.matching_search_term(input.search_term) }
   end
+
+  def with_household_tasks(scope)
+    with_filter(scope, :household_tasks) do
+      return scope.all unless input.household_tasks&.include?('ANNUAL_DUE')
+
+      start_date = Arel.sql <<~SQL
```
Comment (gigxz) [inline drivers/hmis/app/models/hmis/filter/enrollment_filter.rb:66]:
```suggestion
      # Start of the 60-day window during which this year's Annual should be performed. (Anniversary - 30 days)
      start_date = Arel.sql <<~SQL
```

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb [hunk 7]
```diff
@@ -49,4 +50,52 @@ def with_project_types(scope)
...
+        make_date(
+          extract(year from current_date)::integer,
+          extract(month from "EntryDate")::integer,
+          extract(day from "EntryDate")::integer
+        ) - interval '30 days'
+      SQL
+
+      last_start_date = Arel.sql <<~SQL
```
Comment (gigxz) [inline drivers/hmis/app/models/hmis/filter/enrollment_filter.rb:75]:
```suggestion
      # Start of the 60-day window during which _last_ year's Annual should have been performed
      last_start_date = Arel.sql <<~SQL
```

File: drivers/hmis/spec/requests/hmis/project_enrollment_search_spec.rb [hunk 8]
```diff
@@ -86,6 +86,66 @@
...
+              nodes {
+                id
+              }
+            }
+          }
+        }
+      GRAPHQL
+    end
+
+    describe 'household tasks filter' do
```
Comment (gigxz) [inline drivers/hmis/spec/requests/hmis/project_enrollment_search_spec.rb:107]: The tests only cover single-person households. Can you add some where members are in the same households?
That would look like:
```
        # Entered a year ago, annual is due
        e4 = create(:hmis_hud_enrollment, data_source: ds1, project: p1, client: c3, entry_date: Date.today - 1.year)
        # Entered less than a year ago, annual is not due
        create(:hmis_hud_enrollment, data_source: ds1, project: p1, client: c3, entry_date: Date.today - 6.months, household_id: e4.household_id)
```

If there is a 3-person household where only 2 members have completed annuals, the "Annuals Due" filter on the Enrollment table should only include the 1 member that is missing an annual. I think it works this way based on reading the query, but we should have tests to back it. Thanks!

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb [hunk 9]
```diff
@@ -49,4 +50,52 @@ def with_project_types(scope)
...
+            and(cas_t[:data_collection_stage]).eq(5).
+            and(cas_t[:wip]).eq(false),
+          ).join_sources,
+        ).
+        # Earliest entry was more than a year ago
+        where(hh_t[:earliest_entry].lteq(Date.today - 1.year)).
+        # Household is not exited
+        where(hh_t[:latest_exit].eq(nil)).
+        where(
+          # There haven't been any annual assessments yet
```
Comment (gigxz) [inline drivers/hmis/app/models/hmis/filter/enrollment_filter.rb]:
```suggestion
          # This enrollment has not had any annual assessments
```
Maybe clearer? I was confused, at first I thought that this meant there hadn't been any annuals in the _household_.

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb [hunk 10]
```diff
@@ -49,4 +50,52 @@ def with_project_types(scope)
...
+          e_t.
+          join(cas_t, Arel::Nodes::OuterJoin).
+          on(
+            cas_t[:enrollment_id].eq(e_t[:enrollment_id]).
+            and(cas_t[:data_collection_stage]).eq(5).
+            and(cas_t[:wip]).eq(false),
+          ).join_sources,
+        ).
+        # Earliest entry was more than a year ago
+        where(hh_t[:earliest_entry].lteq(Date.today - 1.year)).
```
Comment (gigxz) [inline drivers/hmis/app/models/hmis/filter/enrollment_filter.rb]: I think this should probably be:
```suggestion
        where(hh_t[:earliest_entry].lteq(Date.today - 11.months)).
```

If an enrollment starts on January 1st, their annual "due period" is Dec 1-Jan 31 (roughly).

With the current code, we wouldn't start showing that it was "due" until Jan 1. It seems more helpful to start including the enrollment at the beginning of the "due period" (Dec 1). If the client exits in December, then an annual is/was not necessary. However its probably still useful to include it in this list, since they may remain enrolled.

**Review by gigxz — COMMENTED** [review]
Looking good!! Comments inline

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb (hunk 5 above)
Comment (ZachCMP) [reply drivers/hmis/app/models/hmis/filter/enrollment_filter.rb:17]: I am actually not able to reproduce that if I put in 3/29/2022 for an entry date, in the app or in tests. I switched to using `1 month` as the interval in hopes that that might help. However, if I had to guess I think it'd be from the `make_date(...)` piece, as it would be possible to turn an initial date of 2/29/2020 into 2/29/2023 there. Still, setting a date of 2/29/2020 for entry date doesn't crash anything for me. Maybe we can look at it together?

File: drivers/hmis/spec/requests/hmis/project_enrollment_search_spec.rb (hunk 8 above)
Comment (ZachCMP) [reply drivers/hmis/spec/requests/hmis/project_enrollment_search_spec.rb:107]: Done

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb (hunk 5 above)
Comment (ZachCMP) [reply drivers/hmis/app/models/hmis/filter/enrollment_filter.rb:17]: @gigxz Okay, so I think I might have found a solution. Instead of creating the date directly, I've made it create a date for the first of the month, then add an interval of days to get it to the right day of the month. This might mean that a leap year date of 2/29 will be "due" a day before or after the actual 30 day window on later non-leap years, but that seems like it's maybe okay? I also switched back to `30 days` instead of `1 month` for the interval. Take a look again and see if you still have this problem when you pull down the latest code.

Comment (gigxz) [reply drivers/hmis/app/models/hmis/filter/enrollment_filter.rb:17]: Okay I'll try it out! Thanks!  Its totally fine if leap year dates are off by one.

Comment (gigxz) [reply drivers/hmis/app/models/hmis/filter/enrollment_filter.rb:17]: Alright I'm not getting errors anymore, thought I'm not totally sure I understand why that is. I tried out several configurations including acting like this year is a leap year (2024), and it seemed to work

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb [hunk 11]
```diff
@@ -49,4 +50,54 @@ def with_project_types(scope)
...
+      last_start_date = Arel.sql <<~SQL
+        make_date(
+          extract(year from current_date)::integer,
+          extract(month from "EntryDate")::integer,
+          1
+        ) - interval '30 days' + "interval"((extract(day from "EntryDate")::integer - 1) || ' days') - interval '1 year'
+      SQL
+
+      # SQL-ized version of the logic here: drivers/hmis/app/models/hmis/reminders/reminder_generator.rb#annual_assessment_reminder
+      scope = scope.
```
Comment (gigxz) [inline drivers/hmis/app/models/hmis/filter/enrollment_filter.rb]: Looks like this snuck in accidentally

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb [hunk 12]
```diff
@@ -49,4 +50,54 @@ def with_project_types(scope)
...
+    with_filter(scope, :household_tasks) do
+      return scope.all unless input.household_tasks&.include?('ANNUAL_DUE')
+
+      # Start of the 60-day window during which this year's Annual should be performed. (Anniversary - 30 days)
+      start_date = Arel.sql <<~SQL
+        make_date(
+          extract(year from current_date)::integer,
+          extract(month from "EntryDate")::integer,
+          1
+        ) - interval '30 days' + "interval"((extract(day from "EntryDate")::integer - 1) || ' days')
```
Comment (gigxz) [inline drivers/hmis/app/models/hmis/filter/enrollment_filter.rb:71]: Can you add a comment here about leap year? thanks!

**Review by gigxz — CHANGES_REQUESTED** [review]
looks good to me! just need to fix it so the filter is actually being applied

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb (hunk 11 above)
Comment (ZachCMP) [reply drivers/hmis/app/models/hmis/filter/enrollment_filter.rb]: Should be fixed now

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb (hunk 12 above)
Comment (ZachCMP) [reply drivers/hmis/app/models/hmis/filter/enrollment_filter.rb:71]: Yup done. It's a bit wordy, so let me know if you want me to change it to be more concise

File: drivers/hmis/app/models/hmis/filter/enrollment_filter.rb (hunk 5 above)
Comment (ZachCMP) [reply drivers/hmis/app/models/hmis/filter/enrollment_filter.rb:17]: So I think the issue was that `make_date` was creating an invalid date. So 2/29/2020 is a valid date, but if we use `make_date` to shift the year, then we end up with 2/29/2023, which is not valid. I don't know why `make_date` wasn't erroring out for me in these cases, but it's the only thing I could imagine was creating invalid dates for you and causing the errors.

So to deal with that, we just use `make_date` to create a date on the first day of the anniversary month (e.g. `2/1/2023`), then extract the day of month from the entry date and add it back to the date minus one day. That should result in the same month and day for any other date (e.g. 4/22/2020 -> 4/1/2023 + 21 days == 4/22/2023). For 2/29 it should result in 2/29 on leap years, and 3/1 on non-leap years.

This avoids the possibility of creating an invalid 2/29 date on on a non-leap year, assuming postgres's date math is correct, which I have to assume it generally should be.

Comment (gigxz) [reply drivers/hmis/app/models/hmis/filter/enrollment_filter.rb:17]: Ah ok that makes sense. I thought the problem was the `interval`, but I get it now. Thanks for the comment too!

**Review by gigxz — APPROVED** [review]
looks good!!  I think users are going to be happy about this one 🙂
